ompi/request/req_test.c

```diff
@@ -224,13 +229,25 @@ int ompi_request_default_test_all(
             return MPI_ERR_PROC_FAILED_PENDING;
         }
 #endif /* OPAL_ENABLE_FT_MPI */
+#if OPAL_ENABLE_PROGRESS_THREADS == 0
+        else if (0 == do_it_once) {
+            ++do_it_once;
+            if (0 == opal_progress()) {
+                /* nothing happened, giving up */
+                break;
+            }
+            /* continue walking the list, test the current element again */
+            goto recheck_request_status;
+        }
+#endif /* OPAL_ENABLE_PROGRESS_THREADS */
+        else {
+            /* short-circuit */
+            break;
+        }
```

Comment:
I think you could have made the code more readable, by only increasing the loop index when the request is completed or in active. This way you would not need the label and goto, and the logic of redoing the same loop index after the opal_progress would have been more natural.

Also adding a continue in the case when you increase the loop index would have removed all the "else" in front of the remaining branches.